### PR TITLE
add construct_color_bar for glyphs

### LINKED ISSUE #12736: [FEATURE] Add method to construct a ColorBar
As discussed with @bryevdv, we don't have a simple way to create a ColorBar when using linear_cmap() or log_cmap() (i.e. without explicitly importing and defining a ColorBar object).

The following code works, but it would be good to have an explicit method for this, similar to the [construct method for the contour renderer](https://github.com/bokeh/bokeh/blob/0e72ad3c2f118e0a6d5bb8db6ed048aa07696de0/src/bokeh/models/renderers/contour_renderer.py#L100).

```
r = plot.circle(x, y, color=linear_cmap('y', 'Turbo256', min(y), max(y)), size=10)
color_bar = ColorBar(color_mapper=r.glyph.fill_color.transform)
mapper_plot.add_layout(obj=color_bar, place="right")
```

Optimally, this should work like this:

```
r = plot.circle(...)
color_bar = r.construct_color_bar()
```

The `construct_color_bar` method would take some arguments, including an argument to define if the color_bar should use the mapper associated with a glyph's line or fill color (default to fill color). There also needs to be an error check to verify that the fill_color or line_color is a transform with a colormapper.

### PR DESCRIPTION
**Pull request by bryevdv**
This PR adds a `construct_color_bar` method to `GlyphRenderer` to mirror and be consistent with the one added to the contour renderer. 

@bokeh/core please have a quick glance. If there are no concerns with the proposed API, I will go on to add tests and update relevant examples and documentation. We could potentially add a little more error checking as well (currently will see bare AttributeErrors if asking for "fill" colorbar on a glyph like`Text` with no `fill_color`, but that also seems like a fairly rare corner case)

EDIT: This PR also fixes the "bad column name" validation on glyph renderers. The old code was not straightforward, had no tests, and AFAICT it has actually been completely broken for some time. I added tests for it as well. 

- [x] issues: fixes #12736
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by ianthomas23** [comment]
LGTM so far.

**Comment by bryevdv** [comment]
👍 absent other comments, I will plan to finish this up this weekend.

**Comment by bryevdv** [comment]
This is ready for technical review. @tcmetzger if you have any thoughts about narrative docs, I'm happy to add a small section.

File: src/bokeh/models/renderers/glyph_renderer.py [hunk 1]
```diff
@@ -148,6 +160,41 @@ def add_decoration(self, marking: Marking, node: Literal["start", "middle", "end
...
+            fill_color = self.glyph.fill_color
+            if not (isinstance(fill_color, Field) and isinstance(fill_color.transform, ColorMapper)):
+                raise ValueError("construct_color_bar expects fill_color to be a field with a colormapper transform")
+            return ColorBar(color_mapper=fill_color.transform, **kwargs)
+
+        elif visual == "line":
+            line_color = self.glyph.line_color
+            if not (isinstance(line_color, Field) and isinstance(line_color.transform, ColorMapper)):
+                raise ValueError("construct_color_bar expects line_color to be a field with a colormapper transform")
+            return ColorBar(color_mapper=line_color.transform, **kwargs)
```
Comment (bryevdv) [inline src/bokeh/models/renderers/glyph_renderer.py:204]: Technically these error checks do not cover every contingency. A user could start with a valid mapper then later replace with a plain column name or fixed value, and that would only blow up in the browser. We'd need a serialization validation check to catch that. But I've never seen anyone do that sort of thing, so I'm not too concerned.

**Comment by tcmetzger** [comment]
@bryevdv Thanks for working on this!

My first idea would be to update https://docs.bokeh.org/en/latest/docs/user_guide/basic/annotations.html#color-bars to use the new method instead of the `ColorBar` object. However, will this new method work with this `image` example? If not, we could use a different example here (e.g. a modified version of the hex_tile example in https://docs.bokeh.org/en/latest/docs/user_guide/basic/data.html#client-side-color-mapping located in `examples/basic/data/linear_cmap.py`) - the `image` example could live in the examples then!

We should also update the example at the end of https://docs.bokeh.org/en/latest/docs/user_guide/basic/data.html#client-side-color-mapping (`examples/basic/data/linear_cmap_colorbar.py`)

**Comment by bryevdv** [comment]
@tcmetzger Making it work on `Image` would be a little tricky. As it is, the function expects that the glyph renderer's main glyph has `fill` or `line` visual properties, with a colomapping transform as the value. But that's not the case for `Image`, which has a dedicated property to  hold the bare color mapper for the image data. 

We can certainly make a function that "does the right thing" for images, and I think perhaps we should. But now the API is not quite right, since the `visual` parameter currently only accepts `"fill"` and `"line"`. Some options:

1. Have `construct_color_bar` just ignore the value of `visual` and "do the right thing" when `self.Glyph` is an `Image`. This would be pretty sloppy. 
1. Add a new `"auto"` (or whatever) value for `visual` and make it the default. The meaning of `"auto"` would vary according to the type of `self.glyph`, i.e. use `"fill"` on glyphs that have `"fill"`, use `color_mapper` for `Image`, etc. This is a little loose, but not as sloppy as the first option. 

   We could also consider renaming `visual` to something more generic, and then adding it to contour renderer's version for consistency as well. 
1. Move `construct_color_bar` to the glyph, instead of the glyph render. Then each glyph type can properly define itself how to construct color bars for itself. So for users:
    ```python
    r = p.some_glyph(....)
    r.glyph.construct_color_bar
    ```
    Two issues with this: 1) it's inconsistent with what we already did for contour renderers, with `construct_color_bar` on the renderer, and 2) there's even much less exposure for users to knowing or doing anything with `r.glyph` 
1. Get rid of the `visual` parameter entirely, and `construct_color_bar` just does one thing based on glyph type (i.e. always "fill" for glyphs that have "fill", always "line" for glyphs that only have "line", always "color_mapper" for image... text could use text color, that's a bonus). This (no dedicated args) is already consistent with the contour renderer function as-is. 

   If users need different behaviour, that's when they can create a `Colorbar` manually.
1. Something else entirely?

Thoughts @bokeh/core 

I'm personally inclined to go with option 4) to keep users centered on the glyph renderer in a simple way, but if there is another way to accomplish that, I am all ears.

**Comment by bryevdv** [comment]
I've gone ahead and implemented option 4) which seems like the clear best option under the constraints. @tcmetzger I think all the relevant examples are updated.

File: src/bokeh/models/renderers/glyph_renderer.py [hunk 2]
```diff
@@ -148,6 +160,61 @@ def add_decoration(self, marking: Marking, node: Literal["start", "middle", "end
...
+
+        The function will check for a color mapper on an appropriate property
+        of the GlyphRenderer's main glyph, in this order:
+
+        * ``fill_color.transform`` for FillGlyph
+        * ``line_color.transform`` for LineGlyph
+        * ``text_color.transform`` for TextGlyph
+        * ``color_mapper`` for Image
+
+        In general, the function will "do th right thing" based on glyph type.
```
Comment (tcmetzger) [inline src/bokeh/models/renderers/glyph_renderer.py]:
```suggestion
        In general, the function will "do the right thing" based on glyph type.
```

File: docs/bokeh/source/docs/user_guide/basic/annotations.rst [hunk 3]
```diff
@@ -215,8 +215,20 @@ See |interactive legends| in the user guide for more information and examples.
 Color bars
```
Comment (tcmetzger) [inline docs/bokeh/source/docs/user_guide/basic/annotations.rst:215]: How about we first talk about the `construct_color_bar` as the general way to add a color bar, and then say something like "And you also have the option to create a ColorBar object etc."? 
In the Titles and Legends sections that precede the ColorBar section, we also focus on the plotting interface method first - and the updated colorbar_log.py example uses `construct_color_bar`!

Comment (bryevdv) [reply docs/bokeh/source/docs/user_guide/basic/annotations.rst:215]: I personally prefer this order, since I think its important to establish that the color bar is driven by a color mapper from the start. I also think the example using `construct_color_bar` coming immediately after the introduction of `construct_color_bar` is natural. But if you would like to push changes to this PR I will demur.

File: src/bokeh/models/renderers/glyph_renderer.py [hunk 4]
```diff
@@ -148,6 +160,61 @@ def add_decoration(self, marking: Marking, node: Literal["start", "middle", "end
...
+        The function will check for a color mapper on an appropriate property
+        of the GlyphRenderer's main glyph, in this order:
+
+        * ``fill_color.transform`` for FillGlyph
+        * ``line_color.transform`` for LineGlyph
+        * ``text_color.transform`` for TextGlyph
+        * ``color_mapper`` for Image
+
+        In general, the function will "do the right thing" based on glyph type.
+        If different behavior is needed, ColorBars can be consructed by hand.
```
Comment (tcmetzger) [inline src/bokeh/models/renderers/glyph_renderer.py]:
```suggestion
        If different behavior is needed, ColorBars can be constructed by hand.
```

**Review by tcmetzger — APPROVED** [review]
LGTM, just one random typo!

**Review by ianthomas23 — APPROVED** [review]
